### ISSUE TITLE
metrics calculation problem while test

### ISSUE THREAD

**Issue by hktxt** [qubvel/segmentation_models.pytorch#207]
I follow this [example](https://github.com/qubvel/segmentation_models.pytorch/blob/master/examples/cars%20segmentation%20(camvid).ipynb) to train my data. 
I found a problem.
![捕获](https://user-images.githubusercontent.com/16555750/81900197-f3027100-95ee-11ea-9085-3c4f8e343296.PNG)

above is the loss and iou score while trian/valid. notice that best iou ~0.4, loss: ~1.4.
![捕获](https://user-images.githubusercontent.com/16555750/81900490-791eb780-95ef-11ea-98f2-39793a5c5987.PNG)
and I checked the log, best model saved @ epoch 27. no better one saved after that.

However, when I load the best model and test it on the same validset, it gives a better socre which is a unreliable one.
![捕获](https://user-images.githubusercontent.com/16555750/81900842-327d8d00-95f0-11ea-97cc-d663121d76ae.PNG)
in the hole trianing procedure, the iou never get a score over 0.5, where does 0.66 come from? I comformed that I use the trained model and the same vaildset. It should give the score where the model saved.
did I made any mistake?
thanks for your help.

**Comment by JulienMaille** [comment]
Your progressbars show 23/23 for valid during training and 46/46 when resting afterward. So you may want to check you run your test on the same dataset.
Could also be a different batch size

**Comment by hktxt** [comment]
> Your progressbars show 23/23 for valid during training and 46/46 when resting afterward. So you may want to check you run your test on the same dataset.
> Could also be a different batch size

same set, just different batch size.

**Comment by hktxt** [comment]
same set, just different batch size causes the isuess. But it's ok if use the same batch size. But which score should trust?

**Comment by hktxt** [comment]
It's me again.
same batch_size gives same score. seems ok. 
However, change a different batch size gives another score.
I tested on toy data and trained 10 eopches.
batch size of 2, while valid and test give same score. 
test score matchs the one where the best saved. However, when I change the bs to 1 then test, gives 
a diffferent score.
![捕获24](https://user-images.githubusercontent.com/16555750/82114567-187ab080-9790-11ea-94cb-05cee7ccbc22.PNG)
![捕获25](https://user-images.githubusercontent.com/16555750/82114570-19abdd80-9790-11ea-941e-6d794e482416.PNG)
best model saved @ epoch 9, IOU- 0.4163
test with the same bs:
![捕获26](https://user-images.githubusercontent.com/16555750/82114572-1a447400-9790-11ea-8142-56357ffc2cc2.PNG)
match. test with bs=1:
![捕获27](https://user-images.githubusercontent.com/16555750/82114573-1add0a80-9790-11ea-8997-95a29ca9fa34.PNG)
a different higher score!
It's a bug need to fix.
I hope I made this clear.

**Comment by JulienMaille** [comment]
I suppose that in your dataset you have images where your class is not present. I have noticed that this will strongly affect the IoU.
The thing is that IoU(whole data) != sum(IoU(each image))
The difference is small if your classes are consistently found in all images, but not when absent in some images
